spring-kafka/src/main/java/org/springframework/kafka/core/DefaultKafkaProducerFactory.java

```diff
@@ -846,7 +861,20 @@ protected Producer<K, V> createTransactionalProducer(String txIdPrefix) {
 			}
 		}
 		if (cachedProducer == null) {
-			return doCreateTxProducer(txIdPrefix, "" + this.transactionIdSuffix.getAndIncrement(), this::cacheReturner);
+			String suffix;
+			if (this.maxCache > 0) {
+				BlockingQueue<String> suffixQueue = getSuffixCache(txIdPrefix, this.maxCache);
+				Assert.notNull(suffixQueue,
+						() -> "No suffix cache found for " + txIdPrefix + ", max cache" + this.maxCache);
+				suffix = suffixQueue.poll();
+				if (suffix == null) {
+					throw new NoProducerAvailableException("No transaction producer available for " + txIdPrefix);
+				}
+			}
+			else {
+				suffix = String.valueOf(this.transactionIdSuffix.getAndIncrement());
```

Comment:
Nice work, and thanks for your effort! However, I think just incrementing may not be sufficient. The goal of [GH-2852](https://github.com/spring-projects/spring-kafka/issues/2852) is to allow users to set how `transactionIds` are issued. For example, with a topic having 9 partitions, 3 replicas of my service (each with 3 producers), and limits on 9 `txIds`, restarting my service becomes a bit of a pain. It takes time to distribute transaction IDs correctly(without getting `Fenced` all the time), so instead of that, I would prefer to use external storage as a source of truth.
I suggest introducing a new interface for generating `transactionIdSuffix` with a default increment implementation. WDYTH?

Reply:
> I suggest introducing a new interface for generating transactionIdSuffix with a default increment implementation.
I like the proposal, but there are some scenarios to consider.

Because of maxCache can be modified dynamically, if we didn't set maxCache to begin with
, maybe suffix{0-9} was used. The suffix needs a fixed value to avoid `transaction.id` conflicts.

On the other hand, when it comes to reset(), some transaction ids are in use, we can't recycle them until they close.

The new interface may need to be restricted to KafkaTemplate scenarios where the listenerContainer is reset but some transaction ids are still unavailable, causing conumser to report an error.

Reply:
The first service startup suffix increments from 0 and the second service startup increments from 9?

Reply:
> Via the new interface, does the first instance of the service increment transactionIdSuffix from 0, and the second instance increments it from 3?

It actually depends on the implementation. It could distribute `transactionIdSuffix` between replicas as described, or it could provide the first free suffix from the `source of truth storage`. However, the default implementation remains the same as it is now.

> At the same time, we need to monitor the generated transaction.id. For example, provide a method to return the current transactionIdSuffix or publish an event carrying the current transactionIdSuffix at DefaultKafkaProducerFactory.destroy().

Certainly, we need this functionality. Therefore, the interface must include a contract to retrieve the `transactionId` and return it back.

> As an alternative, writing a unique ID in the transactionIdPrefix, such as UUID, is feasible?


In my environment, it's not an option because I have a mask for `transactionIds` like **topic-name-{0..n}**(e.g. mytopic-1).

Reply:
> At the same time, we need to monitor the generated transaction.id. For example, provide a method to return the current transactionIdSuffix or publish an event carrying the current transactionIdSuffix at DefaultKafkaProducerFactory.destroy().

As a supplement,  interface `ProducerFactory.addListener` can subscribe all producer.